// graphql-ws/dist/Binary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Binary {
    constructor(stream, options) {
        this.onStream = stream;
        this.stream = null;
        this.hash = options.hash;
        this.size = options.size;
        if (Binary.isBinary(options)) {
            this.id = options.id;
        }
        else {
            this.id = ++Binary._id;
        }
    }
    static isBinary(obj) {
        return obj && (typeof obj === 'object') && obj.hasOwnProperty('___binary');
    }
    createReadStream(offset = 0) {
        return this.stream = this.onStream(offset);
    }
    getHash() {
        return this.hash;
    }
    getSize() {
        return this.size;
    }
    getId() {
        return this.id;
    }
    getStream() {
        return this.stream;
    }
    equal(binary) {
        return (this === binary) || (this.hash && binary.hash && this.hash === binary.hash);
    }
    clone(source) {
        this.hash = source.hash;
        this.size = source.size;
        this.id = source.id;
        this.onStream = source.onStream;
        this.stream = source.stream;
    }
    toJSON() {
        return {
            ___binary: true,
            id: this.id,
            hash: this.hash,
            size: this.size,
        };
    }
}
Binary._id = 0;
exports.Binary = Binary;
