// graphql-ws/dist/message-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MessageType;
(function (MessageType) {
    MessageType[MessageType["GQL_CONNECTION_INIT"] = 1] = "GQL_CONNECTION_INIT";
    MessageType[MessageType["GQL_CONNECTION_ACK"] = 2] = "GQL_CONNECTION_ACK";
    MessageType[MessageType["GQL_CONNECTION_ERROR"] = 3] = "GQL_CONNECTION_ERROR";
    MessageType[MessageType["GQL_CONNECTION_KEEP_ALIVE"] = 4] = "GQL_CONNECTION_KEEP_ALIVE";
    MessageType[MessageType["GQL_CONNECTION_TERMINATE"] = 5] = "GQL_CONNECTION_TERMINATE";
    MessageType[MessageType["GQL_START"] = 6] = "GQL_START";
    MessageType[MessageType["GQL_DATA"] = 7] = "GQL_DATA";
    MessageType[MessageType["GQL_BINARY"] = 8] = "GQL_BINARY";
    MessageType[MessageType["GQL_BINARY_ACK"] = 9] = "GQL_BINARY_ACK";
    MessageType[MessageType["GQL_BINARY_REQUEST"] = 10] = "GQL_BINARY_REQUEST";
    MessageType[MessageType["GQL_ERROR"] = 11] = "GQL_ERROR";
    MessageType[MessageType["GQL_COMPLETE"] = 12] = "GQL_COMPLETE";
    MessageType[MessageType["GQL_STOP"] = 13] = "GQL_STOP";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
