// typeorm-transactional-cls-hooked/dist/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var cls_hooked_1 = require("cls-hooked");
var typeorm_1 = require("typeorm");
exports.NAMESPACE_NAME = '__typeOrm___cls_hooked_tx_namespace';
var TYPE_ORM_KEY_PREFIX = '__typeOrm__transactionalEntityManager_';
var TYPE_ORM_HOOK_KEY = '__typeOrm__transactionalCommitHooks';
exports.initializeTransactionalContext = function () {
    return cls_hooked_1.getNamespace(exports.NAMESPACE_NAME) || cls_hooked_1.createNamespace(exports.NAMESPACE_NAME);
};
exports.getEntityManagerOrTransactionManager = function (connectionName, entityManager) {
    var context = cls_hooked_1.getNamespace(exports.NAMESPACE_NAME);
    if (context && context.active) {
        var transactionalEntityManager = exports.getEntityManagerForConnection(connectionName, context);
        if (transactionalEntityManager) {
            return transactionalEntityManager;
        }
    }
    return entityManager || typeorm_1.getManager(connectionName);
};
exports.getEntityManagerForConnection = function (connectionName, context) {
    return context.get("" + TYPE_ORM_KEY_PREFIX + connectionName);
};
exports.setEntityManagerForConnection = function (connectionName, context, entityManager) { return context.set("" + TYPE_ORM_KEY_PREFIX + connectionName, entityManager); };
exports.getHookInContext = function (context) {
    return context.get(TYPE_ORM_HOOK_KEY);
};
exports.setHookInContext = function (context, emitter) {
    return context.set(TYPE_ORM_HOOK_KEY, emitter);
};
