// typeorm-transactional-cls-hooked/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var BaseRepository_1 = require("./BaseRepository");
exports.BaseRepository = BaseRepository_1.BaseRepository;
var BaseTreeRepository_1 = require("./BaseTreeRepository");
exports.BaseTreeRepository = BaseTreeRepository_1.BaseTreeRepository;
var common_1 = require("./common");
exports.initializeTransactionalContext = common_1.initializeTransactionalContext;
exports.getEntityManagerOrTransactionManager = common_1.getEntityManagerOrTransactionManager;
var hook_1 = require("./hook");
exports.runOnTransactionCommit = hook_1.runOnTransactionCommit;
exports.runOnTransactionComplete = hook_1.runOnTransactionComplete;
exports.runOnTransactionRollback = hook_1.runOnTransactionRollback;
var patch_typeorm_repository_1 = require("./patch-typeorm-repository");
exports.patchTypeORMRepositoryWithBaseRepository = patch_typeorm_repository_1.patchTypeORMRepositoryWithBaseRepository;
exports.patchTypeORMTreeRepositoryWithBaseTreeRepository = patch_typeorm_repository_1.patchTypeORMTreeRepositoryWithBaseTreeRepository;
var Propagation_1 = require("./Propagation");
exports.Propagation = Propagation_1.Propagation;
var IsolationLevel_1 = require("./IsolationLevel");
exports.IsolationLevel = IsolationLevel_1.IsolationLevel;
var Transactional_1 = require("./Transactional");
exports.Transactional = Transactional_1.Transactional;
__export(require("./TransactionalError"));
