// nestjs-admin/dist/bin/cliAdmin.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var CliAdminModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@nestjs/common");
const typeorm_1 = require("@nestjs/typeorm");
const typeormProxy_1 = require("../src/utils/typeormProxy");
const adminUser_entity_1 = require("../src/adminUser.entity");
const adminUser_service_1 = require("../src/adminUser.service");
class AdminConnectionException extends Error {
    constructor(msg) {
        super(msg);
    }
}
let CliAdminModule = CliAdminModule_1 = class CliAdminModule {
    static getConnectionOptions() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                return yield typeormProxy_1.getConnectionOptions();
            }
            catch (e) {
                throw new AdminConnectionException(`Could not connect to your database to create an admin user. Make sure you have an ormconfig file with a default connection or that your environment variables are set (see https://github.com/typeorm/typeorm/blob/master/docs/using-ormconfig.md).`);
            }
        });
    }
    static create() {
        return __awaiter(this, void 0, void 0, function* () {
            const connectionOptions = yield this.getConnectionOptions();
            const entities = connectionOptions.entities || [];
            return {
                module: CliAdminModule_1,
                imports: [
                    typeorm_1.TypeOrmModule.forRoot(Object.assign({}, connectionOptions, { entities: [...entities, adminUser_entity_1.default] })),
                    typeorm_1.TypeOrmModule.forFeature([adminUser_entity_1.default]),
                ],
            };
        });
    }
};
CliAdminModule = CliAdminModule_1 = __decorate([
    common_1.Module({
        providers: [adminUser_service_1.AdminUserService],
    })
], CliAdminModule);
exports.CliAdminModule = CliAdminModule;
